### ISSUE TITLE
Nvidia shield

### ISSUE THREAD

**Issue by sanchescom** [skyline-emu/skyline#202]
Does it work on Nvidia shield? I am trying to install but without success.

**Comment by bylaws** [comment]
No, Android tv is really limited and offers no APIs to open a directory with SAF

**Comment by alexanderadam** [comment]
Are you sure about that?
Do you have any details in documentation because I wasn't able to find such limitation.
The Android TV devices that I own support everything my other devices support as well.

I thought that Android TV devices only need to fulfil certain metadata, have certain layout support and of course support for non-touch input devices.
There's some summary [in the documentation](https://developer.android.com/training/tv/games) and there's also [a nice checklist](https://developer.android.com/training/tv/publishing/checklist).

NVidia [also has some documentation](https://developer.nvidia.com/android-tv-deployment-checklist) in case that's more suitable to you.

Nvidia also [provides images](https://developer.nvidia.com/gameworksdownload) for their Android TV devices.

Furthermore the claim

>  Most of these run weak 32bit socs

is very misleading. The NVidia Shield is able to run some decent games.

**Comment by jduncanator** [comment]
> For instance the popular NVidia Shield is able to run some decent games and other emulators like Dolphin emulator.

Perhaps because Dolphin aims to emulate a console released 11 years before the Nintendo Switch.

The Shield only ships with 4GB of RAM, which is generally accepted in the Switch emulation community to be insufficient for emulating a console which itself has 4GB of RAM.

**Comment by alexanderadam** [comment]
> The Shield only ships with 4GB of RAM, which is generally accepted in the Switch emulation community to be insufficient for emulating a console which itself has 4GB of RAM.

Well, I'm sure you were already aware of these points but just in case you aren't:

- RAM usage is strongly dependent on the application/game. It would surely be strange if every game would always max out the whole memory (or even swap)
- People are already using Skyline on phones with 4GB RAM
- Also just a little reminder that there are Android TV boxes with more RAM and I guess this will always be the case

So apart from this also_this_ argument being absolutely misleading, it isn't event relevant for this issue.

Because I mentioned before this isn't about resources but about gaining _compatibility_.
And Android TV compatibility is _not_ related to memory or CPU constraints but mainly about metadata, layouts and controls.
So with these changes people could install and run Skyline on Android TV distributions in the same way they can do it on regular Android devices.
Regardless which specs the hardware have.

I mentioned the relevant documentation [in my last comment](https://github.com/skyline-emu/skyline/issues/202#issuecomment-1186133934) in case this got missing somehow.

Skyline won't run slower or eat more memory just because it has Android TV compatibility.

**Comment by PixelyIon** [comment]
**TLDR:** We will fix devices that don't properly support SAF (Such as the Oculus Quest 2 or most Android TV devices) eventually, certain games can run at that point. This is fundamentally not a priority since the vast majority of Android TV devices are extremely underpowered, the Shield TV is the single exception but still won't be able to run major Switch titles due to the aforementioned RAM concerns.

> Are you sure about that?
> Do you have any details in documentation because I wasn't able to find such limitation and it is the first time I'm hearing about it.

You can try running the [`ACTION_OPEN_DOCUMENT_TREE` Intent](https://developer.android.com/reference/android/content/Intent#ACTION_OPEN_DOCUMENT_TREE) on Android TV devices right now and see that it'll fail, it should open up the directory picker but the vendors don't implement it as the [Android CTS](https://source.android.com/compatibility/cts) doesn't check for it on Android TV devices.

> For instance the popular NVidia Shield is able to run some decent games and other emulators like [Dolphin emulator](https://www.youtube.com/watch?v=DSr8xu9vvuk).

The issue is a lack of *any* other instance of good SoCs in Android TV except the Tegra X1, Dolphin is a special example since Android support on it was brought up on NV devices to begin with and Dolphin emulates the Wii/GC which can't be compared to the Switch in any regard when it comes to the specifications required especially concerning RAM.

> Therefore I would be curious why every other emulator runs perfectly fine on Android TV while Skyline wouldn't.

I've talked to other emulator developers about this and it's because of depending on legacy non-SAF support. Skyline being a relatively new emulator had SAF implemented [as far back as 2019](https://github.com/skyline-emu/skyline/commit/e4dc602f4d2d38686581365602262370f71d156b). SAF is the future of Android FS access and [non-SAF FS access has been deprecated](https://developer.android.com/about/versions/11/privacy/storage#scoped-storage).

Once you implement SAF, there's no real way to do things properly and we're the only emulator I'm aware of that has a `DocumentProvider` which is the only way SAF intended access to common directories and will allow importing ROMs from it. As long as it's not broken on Android TV again, that's the only solution that should universally work even on devices without proper SAF intent support.

I should reiterate that this is Android's fault for not making the CTS extensive enough and the vendor's fault for not implementing SAF even if the CTS doesn't check for it. Absolutely none of this is the fault of the application, we're just left to deal with the broken and incomplete APIs that have been provided to us.

> Skyline won't run slower or eat more memory just because it has Android TV compatibility.

It will take up valuable developer time, it doesn't take a genius to be able to tell that we have far more important things to work on than support for a single Android TV device. There's also the case of developer interest which you can't take for granted, if no developers are particularly interested in something then it's not happening. Our development is user-driven and there's just not enough demand for this, especially when weighed up to the demand for other things such as larger game support and better performance. [No amount of linking to random guides will change that.](https://github.com/skyline-emu/skyline/issues/202#issuecomment-1186133934)

**Comment by alexanderadam** [comment]
> We will fix devices that don't properly support SAF (Such as the Oculus Quest 2 or most Android TV devices) eventually, certain games can run at that point.

[According to a user called mitvboxfactory](https://www.nvidia.com/en-us/geforce/forums/shield-tv/9/494556/storage-access-framework-picker-ui-doesnt-work-wit/)

> SAF functionality works fine on SHIELD EXPERIENCE 9, but the picker UI does not work with a shield controller, I have to use a mouse to select item in the left side

So I guess apart from the Shield controller UX SAF doesn't seem to be general out of scope here. I just searched for shield and SAF :shrug: 

> it doesn't take a genius to be able to tell that we have far more important things to work on than support for a single Android TV device

Of course not. It also doesn't take a genius to reopen a GitHub issue that was closed before because of invalid reasons.
Doesn't it? :wink: 

Nobody here asked anyone to implement this _right now_ (_I hope nobody is doing this for other issues either_).
Folks are just asking whether Android TV support would be a realistic request.
And open & transparent GitHub issues are pretty good in tracking them.

Hence it would be great to reopen an Android TV issue (or create a new one if you didn't like the former), if no one claims any more that Android TV devices can't have a good CPU or amount of RAM that's comparable to the mobile devices where Skyline is currently already running. :wink: 

Because 

> [There is no way in hell any android tv is capable of running skyline.](https://github.com/skyline-emu/skyline/issues/277#issuecomment-966314199)

is just a very unlikely claim.

**Comment by PixelyIon** [comment]
> So I guess apart from the user experience of the Shield Controller, SAF doesn't seem to be general out of scope here.

It seems like that's a very new development, if that works then Skyline should just work on the Shield without modification as NV is a good vendor. Which does further devalue reasoning to support devices without proper SAF since vendors can just be less lazy and do it themselves but alas as we're on Android we just can't depend on vendors to release devices that aren't broken.

> Of course not. It also doesn't take a genius to reopen a GitHub issue that was closed before because of invalid reasons.

NV added SAF support so this issue is effectively solved. Regardless, even if it was an issue about Android TV as it isn't something on our roadmap and more of a thing we're willing to do if anyone is interested, it wouldn't remain open.

> Folks are just asking whether Android TV support would be a realistic request.

As far as that's concerned, we already do support the vast majority of flags to provide the optimal experience on Android TV, our UI has been entirely controller navigable for a long time. The only thing that would improve the experience that we enable is [HDMI game mode](https://developer.android.com/reference/android/view/Window#setPreferMinimalPostProcessing(boolean)). We've always set the application to show up as a game application and other manifest flags only end up mattering for the Play Store and don't change the app UX in any way.

The only thing that makes Skyline not work on Android TV is because of vendors shipping incomplete versions of SAF which we need to find ways to work around and I've already described the way we're going to take but it's not something anyone is entirely willing to do, it's just something we *may* do.

I should reiterate that we aren't unreasonable and given there is substantial push from users about it, there's more of a chance that we'll work on it but I'm unsure if that'll be the case.

**Comment by bylaws** [comment]
I should also note that something like a leanback UI is never going to be implemented, it would require a lot of work and upkeep for almost no users. The shield TV is also super weak so the most you'll be able run is hollow knight or something, giving most users an overall bad experience.

If Android TV suddenly became a platform for high end devices however and gained significant popularity we would definitely look into a better experience though.
